packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -68,26 +68,18 @@ const BlockComponent = forwardRef(
 		const { removeBlock, insertDefaultBlock } = useDispatch(
 			'core/block-editor'
 		);
-		const fallbackRef = useRef();
+		const wrapper = useRef();
 		const isAligned = wrapperProps && !! wrapperProps[ 'data-align' ];
-		wrapper = wrapper || fallbackRef;
 
-		// Provide the selected node, or the first and last nodes of a multi-
-		// selection, so it can be used to position the contextual block toolbar.
-		// We only provide what is necessary, and remove the nodes again when they
-		// are no longer selected.
+		// Provide the selected node so it can be used to position the contextual block toolbar.
+		// The wrapper changes when the alignment class changes.
 		useLayoutEffect( () => {
-			if ( isSelected || isFirstMultiSelected || isLastMultiSelected ) {
-				const node = wrapper.current;
-				setBlockNodes( ( nodes ) => ( {
-					...nodes,
-					[ clientId ]: node,
-				} ) );
-				return () => {
-					setBlockNodes( ( nodes ) => omit( nodes, clientId ) );
-				};
-			}
-		}, [ isSelected, isFirstMultiSelected, isLastMultiSelected ] );
+			setBlockNodes( ( nodes ) => ( {
+				...nodes,
+				[ clientId ]: wrapper.current,
+			} ) );
+			return () => setBlockNodes( ( nodes ) => omit( nodes, clientId ) );
+		}, [ isAligned ] );
```

Comment:
The fix here is that the block  node is set now for all blocks (not just selected blocks) and is refreshed only when the  "isAligned" changes because that's the only thing that can affect it here.

Reply:
Why do we need all block nodes for alignment?
I suspect that this will impact loading performance because this blocks layout. Perhaps we can use `useEffect` instead, but I recall having some e2e test problems with that.

Reply:
Generally, I do think storing all reference would be useful, especially since it could replace the "ugly" querying of block nodes in other places. Not sure why we need it in this PR though.

Reply:
I checked the performance impact and I don't think there's one. That said I'd appreciate confirmation since there are fluctuations.

Reply:
> Why do we need all block nodes for alignment?

Assigning or unassigning an alignment rerenders the block entirely and generates a new  "node" for this block. (because there's an extra wrapper needed)

Reply:
Oh, does a component not remount when that happens?

Reply:
no, it doesn't, the element change but the component is still the same.

Reply:
Hm, I wonder if we should update the list on every rerender, if the node is not the same.

Reply:
Or maybe we should store the ref object rather than the element itself?

Reply:
the  node only changes on mount/unmout and on alignment set/unset because the element tree changes at that moment. I tried using a "function as ref" because that's that function is called at the exact moment the ref changes but it doesn't work because that function is called  at "render" time by React causing infinite loops and weird breakage.

What I have now is basically the same thing as "function as ref" but called as an effect (once the render is finished).